src/ruby/puppetserver-lib/puppet/server/http_client.rb

```diff
@@ -44,17 +39,23 @@ def self.settings
     @settings ||= {}
   end
 
-  def initialize(server, port, options = {})
-    options = OPTION_DEFAULTS.merge(options)
+  def initialize(options = {})
+    # NOTE: Unlike Puppet's HTTP client implementation, the Java HttpAsyncClient
+    # does not support retries in the version we are using, so the `retry_limit`
+    # option will be ignored if passed.
+    #
+    # Similarly, `redirect_limit` is not easily configurable, so we just use the default
+    # of 100. The param will be ignored if passed.
+  end
 
-    @server = server
-    @port = port
-    @use_ssl = options[:use_ssl]
-    @protocol = @use_ssl ? "https" : "http"
+  def get(url, headers: {}, params: {}, options: {}, &block)
```

Comment:
I think I lost something here (that isn't used because the old API didn't support it). This should yield the response object like Puppet does if a block is given, in addition to always returning the response object at the end.

Reply:
Though I'm actually a little confused what https://github.com/puppetlabs/puppet/blob/main/lib/puppet/http/client.rb#L133 accomplishes in Puppet's implementation. @joshcooper what is that line for? Do we need it here or does it have something to do with the underlying client implementation? Or am I just spacing on something odd Ruby is doing here?

Reply:
Sorry I misspoke, it's really due to persistent connections. When we make a `get` etc request, we borrow a connection from the pool and return it before the `get` method returns. We need to know when the connection is returned whether it can be reused later or should be discarded. If the `get` method returned a response object, but the body wasn't read until sometime later, then it would be reading from the socket associated with a connection that was back in the pool, or worse handed out to someone else, leading to streams getting crossed. So we need to make sure that the response body is fully consumed before the connection is handed back to the pool, which needs to happen before the `get` method returns.

I think it would be good for puppetserver to keep the same semantics as I believe that's how Apache httpclient behaves also. Note sure I have the apache httpclient version correct, but what I've been calling `drain` is what they call `consume` in the description example: https://hc.apache.org/httpcomponents-client-4.5.x/httpclient/apidocs/org/apache/http/impl/client/package-summary.html#package_description

```
      HttpGet httpGet = new HttpGet("http://targethost/homepage");
      CloseableHttpResponse response = httpclient.execute(httpGet);
      try {
          System.out.println(response.getStatusLine());
          HttpEntity entity = response.getEntity();
          // do something useful with the response body
          // and ensure it is fully consumed
          EntityUtils.consume(entity);
      } finally {
          response.close();
      }
```

If puppetserver kept the same semantics it would make it easier to support persistent connections between puppetserver and puppetdb in the future.

Reply:
I'm also working on a PR to updated the http client yardocs as it's definitely missing details like this.

Reply:
Currently getting a bunch of test failures when I try to add `EntityUtils.consume`. I sort of suspect the fact that we are using the async Apache client under the hood, even for sync clients, might be interfering with it. Need to look into it more to see if what we're already doing is sufficient.

Reply:
So we're definitely not calling `EntityUtils.consume`, but my understanding was that, if the response is set as TEXT (which we force [here](https://github.com/puppetlabs/puppetserver/blob/6.x/src/ruby/puppetserver-lib/puppet/server/http_client.rb#L104)) then we read the response body, attempt to coerce it to the encoding, and close the underlying resource ([here](https://github.com/puppetlabs/clj-http-client/blob/ec78805007db944b0ef7b4e4212d4969d7d8d0a4/src/java/com/puppetlabs/http/client/impl/JavaClient.java#L666)). We do similar for streaming response [here](https://github.com/puppetlabs/clj-http-client/blob/ec78805007db944b0ef7b4e4212d4969d7d8d0a4/src/java/com/puppetlabs/http/client/impl/StreamingAsyncResponseConsumer.java#L49) in our client wrapper. I don't see a safe way to use the non-streaming byte responses in our jvm client library, but we don't expose that in the Ruby client. I'm also not sure if the TEXT processing wouldn't fail to close the resources if there was an encoding error (but I've been unable to verify that).

Reply:
We could move that `if body then close` bit to a finally block maybe to be safe? Not sure if there's a reason it's not like that already...